### ISSUE TITLE
(proposal) Stock Consumption Design

### ISSUE THREAD

**Issue by mbayopanda** [third-culture-software/bhima#1853]
Tracking the stock consumption is very important for the stock management, so here is a proposal about the stock consumption design and implementation.

The idea in this proposal is to have a table named `stock_consumption` which aggregate consumption by period such `period_total`.

### Stock Consumption Table

```sql
CREATE TABLE `stock_consumption` (
  `inventory_uuid`  BINARY(16) NOT NULL,
  `depot_uuid`      BINARY(16) NOT NULL,
  `period_id`       MEDIUMINT(8) NOT NULL,
  `quantity`        INT(11) UNSIGNED DEFAULT NULL,
  PRIMARY KEY (`inventory_uuid`, `depot_uuid`, `period_id`),
  KEY `inventory_uuid` (`inventory_uuid`),
  KEY `depot_uuid` (`depot_uuid`),
  KEY `period_id` (`period_id`),
  FOREIGN KEY (`inventory_uuid`) REFERENCES `inventory` (`uuid`),
  FOREIGN KEY (`depot_uuid`) REFERENCES `depot` (`uuid`),
  FOREIGN KEY (`period_id`) REFERENCES `period` (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

### ReComputeConsumption Procedure

This procedure will recalculate the consumption of an inventory in a given depot. We will call this procedure each time when we make a consumption that we want track, for example when we distribute stock (inventory) to patients.

```sql
CREATE PROCEDURE ReComputeConsumption (
  IN inventory_uuid BINARY(16),
  IN depot_uuid BINARY(16),
  IN period_id MEDIUMINT(8),
  IN quantity INT(11)
) 
BEGIN 
  INSERT INTO `stock_consumption` (inventory_uuid, depot_uuid, period_id, quantity) VALUES 
    (inventory_uuid, depot_uuid, period_id, quantity)
  ON DUPLICATED KEY UPDATE quantity = quantity + VALUES(quantity);
END $$
```

## Report

We can have a basic configurable report which can have as input : 
- depot : Not required if we want to have report for all depots
- inventory : Not required if we want to have report for all inventories
- period_start : Required
- period_end : Required

Output : 
- Inventory
- Depot
- CM : Consommation Mensuelle (Monthly Consumption)
- CMM : Consommation Mensuelle Moyenne (!important)

**Comment by jniles** [comment]
This is excellent English (!) 👍 

I imagine that the `RecomputeConsumption()` procedure will be called a lot, if we have it on every patient consumption - in theory, those should happen as regularly as Cash Payments.  We'll have to do some performance testing for this.

LGTM.

**Comment by mbayopanda** [comment]
Closed by PR #1885
